src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations/AvoidMultipleEnumerations.DataFlowOperationVisitor.cs

```diff
@@ -0,0 +1,140 @@
+            {
+                var value = base.VisitParameterReference(operation, argument);
+                if (!IsDeferredType(operation.Parameter.Type.OriginalDefinition, _wellKnownSymbolsInfo.AdditionalDeferredTypes))
+                {
+                    return value;
+                }
+
+                return CreateAndUpdateAnalysisValue(operation, value);
+            }
+
+            public override GlobalFlowStateDictionaryAnalysisValue VisitLocalReference(ILocalReferenceOperation operation, object? argument)
+            {
+                var value = base.VisitLocalReference(operation, argument);
+                if (!IsDeferredType(operation.Local.Type.OriginalDefinition, _wellKnownSymbolsInfo.AdditionalDeferredTypes))
+                {
+                    return value;
+                }
+
+                return CreateAndUpdateAnalysisValue(operation, value);
+            }
+
+            private GlobalFlowStateDictionaryAnalysisValue CreateAndUpdateAnalysisValue(IOperation operation, GlobalFlowStateDictionaryAnalysisValue defaultValue)
+            {
+                if (!IsOperationEnumeratedByMethodInvocation(operation, _wellKnownSymbolsInfo) && !IsGetEnumeratorOfForEachLoopInvoked(operation))
+                {
+                    return defaultValue;
+                }
+
+                if (!TryGetInvocationEntity(operation, out var invocationEntity))
+                {
+                    return defaultValue;
+                }
+
+                var newValue = CreateAnalysisValue(invocationEntity, operation, defaultValue);
+                UpdateGlobalValue(newValue);
+                return newValue;
+            }
+
+            private bool TryGetInvocationEntity(IOperation operation, [NotNullWhen(true)] out InvocationEntity? invocationEntity)
+            {
+                var pointToAnalysisResult = DataFlowAnalysisContext.PointsToAnalysisResult;
+                invocationEntity = null;
+                if (pointToAnalysisResult is null)
+                {
+                    return false;
+                }
+
+                var result = pointToAnalysisResult[operation.Kind, operation.Syntax];
+                if (result.Kind != PointsToAbstractValueKind.KnownLocations)
+                {
+                    return false;
+                }
+
+                invocationEntity = new InvocationEntity(result.Locations);
+                return true;
+            }
+
+            private void UpdateGlobalValue(GlobalFlowStateDictionaryAnalysisValue value)
+            {
+                if (value.Kind == GlobalFlowStateDictionaryAnalysisValueKind.Known)
+                {
+                    var newState = GlobalFlowStateDictionaryAnalysisValue.Merge(GlobalState, value);
+                    SetAbstractValue(GlobalEntity, newState);
+                }
+            }
+
+            private bool IsGetEnumeratorOfForEachLoopInvoked(IOperation operation)
```

Comment:
Should this helper method also be moved to the subsequent file while has all helper methods? Any reason this method is not a static method?

Reply:
Yep